### ISSUE TITLE
[JENKINS-27329] WorkspaceCleanupThread may delete workspaces of running jobs

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-14#12309]
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/qhartman.png" title="qhartman's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=qhartman">qhartman</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-27329" target="_blank">WorkspaceCleanupThread may delete workspaces of running jobs</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-06-06T09:43:24+00:00
<li><b>votes</b>: 13
<li><b>watchers</b>: 19
<li><b>imported</b>: 2025-11-14
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Linux host, Linux, OSX, Windows, slaves. Jenkins version 1.602.</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=234136&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-234136">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>&gt;&gt; Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154
</pre>
</div></div>
<p><em>side-note</em>: it's a shame those log lines are not time-stamped
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>
</pre>
</details>

</details>
<p>>> Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old<br/>
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>


```
Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154
```
<p><em>side-note</em>: it's a shame those log lines are not time-stamped<br/>
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=234137&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-234137">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Daniel knows about this area, so assigning to him for comment (sorry, Daniel!)</p>
</pre>
</details>

</details>
<p>Daniel knows about this area, so assigning to him for comment (sorry, Daniel!)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bonefish's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=238858&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-238858">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Since <a href="https://issues.jenkins.io/browse/JENKINS-30916" title="workspace being deleted mid-build on slave" class="issue-link" data-issue-key="JENKINS-30916"><del>JENKINS-30916</del></a> has been closed as a duplicate: Here the ticket description only says that workspaces that aren't old are deleted. In fact a workspace can even be deleted while a build using the workspace is in progress. The lines from the system log for such a case:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Okt 13, 2015 3:29:27 AM INFORMATION hudson.slaves.CommandLauncher launch
slave agent launched for BonefishMac-Ubuntu-12.04
Okt 13, 2015 3:31:15 AM INFORMATION hudson.model.AsyncPeriodicWork$1 run
Started Workspace clean-up
Okt 13, 2015 3:31:21 AM INFORMATION hudson.model.Run execute
Bar-Nightly/label=Ubuntu-12.04 #222 main build action completed: FAILURE
</pre>
</div></div>
</pre>
</details>

</details>
<p>Since <a href="https://issues.jenkins.io/browse/JENKINS-30916" title="workspace being deleted mid-build on slave" class="issue-link" data-issue-key="JENKINS-30916"><del>JENKINS-30916</del></a> has been closed as a duplicate: Here the ticket description only says that workspaces that aren't old are deleted. In fact a workspace can even be deleted while a build using the workspace is in progress. The lines from the system log for such a case:</p>


```
Okt 13, 2015 3:29:27 AM INFORMATION hudson.slaves.CommandLauncher launch
slave agent launched for BonefishMac-Ubuntu-12.04
Okt 13, 2015 3:31:15 AM INFORMATION hudson.model.AsyncPeriodicWork$1 run
Started Workspace clean-up
Okt 13, 2015 3:31:21 AM INFORMATION hudson.model.Run execute
Bar-Nightly/label=Ubuntu-12.04 #222 main build action completed: FAILURE
```

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>davida2009's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=283274&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283274">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I experienced this bug last night.  The workspace of a matrix job was deleted while the job was running.</p>

<div class="panel" style="border-width: 1px;"><div class="panelContent">
<p>Jan 12, 2017 9:00:00 PM INFO hudson.triggers.SCMTrigger$Runner run
SCM changes detected in tml_system_level_regression_tests_linux_all_branches_and_trunk » branches/TRY_TML_LEDA_17May2016. Triggering #125
Jan 12, 2017 9:00:00 PM INFO hudson.triggers.SCMTrigger$Runner run
SCM changes detected in tml_system_level_regression_tests_linux_all_branches_and_trunk » trunk. Triggering #153
Jan 12, 2017 10:35:00 PM INFO hudson.triggers.SCMTrigger$Runner run
SCM changes detected in Regression_test_TestParams_VisualStudio. Triggering #403
Jan 12, 2017 10:41:41 PM INFO hudson.model.Run execute
Regression_test_TestParams_VisualStudio #403 main build action completed: SUCCESS
Jan 12, 2017 11:03:20 PM INFO hudson.model.AsyncPeriodicWork$1 run
Started Workspace clean-up
Jan 12, 2017 11:03:48 PM INFO hudson.model.AsyncPeriodicWork$1 run
Finished Workspace clean-up. 27,337 ms
Jan 12, 2017 11:04:51 PM INFO hudson.model.Run execute
tml_system_level_regression_tests_linux_all_branches_and_trunk/trunk #153 main build action completed: FAILURE</p>
</div></div>

<p>We are running Jenkins 2.40 with Multi-Branch Project Plugin 0.3.</p>
</pre>
</details>

</details>
<p>I experienced this bug last night.  The workspace of a matrix job was deleted while the job was running.</p>



<table><tr><td><p>Jan 12, 2017 9:00:00 PM INFO hudson.triggers.SCMTrigger$Runner run<br/>
SCM changes detected in tml_system_level_regression_tests_linux_all_branches_and_trunk » branches/TRY_TML_LEDA_17May2016. Triggering #125<br/>
Jan 12, 2017 9:00:00 PM INFO hudson.triggers.SCMTrigger$Runner run<br/>
SCM changes detected in tml_system_level_regression_tests_linux_all_branches_and_trunk » trunk. Triggering #153<br/>
Jan 12, 2017 10:35:00 PM INFO hudson.triggers.SCMTrigger$Runner run<br/>
SCM changes detected in Regression_test_TestParams_VisualStudio. Triggering #403<br/>
Jan 12, 2017 10:41:41 PM INFO hudson.model.Run execute<br/>
Regression_test_TestParams_VisualStudio #403 main build action completed: SUCCESS<br/>
Jan 12, 2017 11:03:20 PM INFO hudson.model.AsyncPeriodicWork$1 run<br/>
Started Workspace clean-up<br/>
Jan 12, 2017 11:03:48 PM INFO hudson.model.AsyncPeriodicWork$1 run<br/>
Finished Workspace clean-up. 27,337 ms<br/>
Jan 12, 2017 11:04:51 PM INFO hudson.model.Run execute<br/>
tml_system_level_regression_tests_linux_all_branches_and_trunk/trunk #153 main build action completed: FAILURE</p></td></tr></table>


<p>We are running Jenkins 2.40 with Multi-Branch Project Plugin 0.3.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>davida2009's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=283275&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283275">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Please could this issue be assigned to someone?</p>
</pre>
</details>

</details>
<p>Please could this issue be assigned to someone?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>davida2009's <a href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=283277&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283277">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Matthew, please forgive my ignorance but where do I set hudson.model.WorkspaceCleanupThread.disabled=true?</p>
</pre>
</details>

</details>
<p>Matthew, please forgive my ignorance but where do I set hudson.model.WorkspaceCleanupThread.disabled=true?</p>
